### DISCUSSION TITLE
Does choice guarantee order?

### DISCUSSION THREAD

**Discussion by baanan** [zesterer/chumsky#133] [Q&A] [answered]
In the documentation for `or`, it says

> If both parsers succeed, the output of the first parser is guaranteed to be prioritized over the output of the second.

Does this also hold true for `choice`?

**Comment by zesterer** [comment, marked as answer]
Yes, this also holds for `choice`. Chumsky is a PEG parser, so ambiguity is resolved by simply using the order of patterns as the tie-breaking mechanism.

As mentioned in the docs, `choice` has a few tradeoff compared to `or` chains:

- It's significantly faster to compile (I'm hoping that a future version of Rust will be able to memoise trait solving to fix this)
- Its error selection is very slightly weaker (but still sufficient for 90% of use-cases)

**Reply by baanan** [reply]
ok, thanks
